In [3]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 1000)

In [4]:
# ============================
# Load the datasets
# ============================

users = pd.read_csv("../data/raw/users.csv")

flights = pd.read_csv("../data/raw/flights.csv")

hotels = pd.read_csv("../data/raw/hotels.csv")

In [5]:
users.head(), flights.head(), hotels.head()

(   code company             name  gender  age
 0     0    4You        Roy Braun    male   21
 1     1    4You   Joseph Holsten    male   37
 2     2    4You    Wilma Mcinnis  female   48
 3     3    4You     Paula Daniel  female   23
 4     4    4You  Patricia Carson  female   44,
    travelCode  userCode                from                  to  flightType    price  time  distance       agency        date
 0           0         0         Recife (PE)  Florianopolis (SC)  firstClass  1434.38  1.76    676.53  FlyingDrops  09/26/2019
 1           0         0  Florianopolis (SC)         Recife (PE)  firstClass  1292.29  1.76    676.53  FlyingDrops  09/30/2019
 2           1         0       Brasilia (DF)  Florianopolis (SC)  firstClass  1487.52  1.66    637.56      CloudFy  10/03/2019
 3           1         0  Florianopolis (SC)       Brasilia (DF)  firstClass  1127.36  1.66    637.56      CloudFy  10/04/2019
 4           2         0        Aracaju (SE)       Salvador (BH)  firstClass  1684

In [6]:
print("Users Shape :", users.shape)

print("Flights Shape :", flights.shape)

print("Hotels Shape :", hotels.shape)

Users Shape : (1340, 5)
Flights Shape : (271888, 10)
Hotels Shape : (40552, 8)


In [7]:
users.columns

Index(['code', 'company', 'name', 'gender', 'age'], dtype='str')

In [8]:
flights.columns

Index(['travelCode', 'userCode', 'from', 'to', 'flightType', 'price', 'time', 'distance', 'agency', 'date'], dtype='str')

In [9]:
hotels.columns

Index(['travelCode', 'userCode', 'name', 'place', 'days', 'price', 'total', 'date'], dtype='str')

In [10]:
users.info()

<class 'pandas.DataFrame'>
RangeIndex: 1340 entries, 0 to 1339
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   code     1340 non-null   int64
 1   company  1340 non-null   str  
 2   name     1340 non-null   str  
 3   gender   1340 non-null   str  
 4   age      1340 non-null   int64
dtypes: int64(2), str(3)
memory usage: 52.5 KB


In [11]:
flights.info()

<class 'pandas.DataFrame'>
RangeIndex: 271888 entries, 0 to 271887
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   travelCode  271888 non-null  int64  
 1   userCode    271888 non-null  int64  
 2   from        271888 non-null  str    
 3   to          271888 non-null  str    
 4   flightType  271888 non-null  str    
 5   price       271888 non-null  float64
 6   time        271888 non-null  float64
 7   distance    271888 non-null  float64
 8   agency      271888 non-null  str    
 9   date        271888 non-null  str    
dtypes: float64(3), int64(2), str(5)
memory usage: 20.7 MB


In [12]:
hotels.info()

<class 'pandas.DataFrame'>
RangeIndex: 40552 entries, 0 to 40551
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   travelCode  40552 non-null  int64  
 1   userCode    40552 non-null  int64  
 2   name        40552 non-null  str    
 3   place       40552 non-null  str    
 4   days        40552 non-null  int64  
 5   price       40552 non-null  float64
 6   total       40552 non-null  float64
 7   date        40552 non-null  str    
dtypes: float64(2), int64(3), str(3)
memory usage: 2.5 MB


In [13]:
# Check missing values in all datasets

print("Users Dataset - Missing Values")
print(users.isnull().sum())

print("\nFlights Dataset - Missing Values")
print(flights.isnull().sum())

print("\nHotels Dataset - Missing Values")
print(hotels.isnull().sum())

Users Dataset - Missing Values
code       0
company    0
name       0
gender     0
age        0
dtype: int64

Flights Dataset - Missing Values
travelCode    0
userCode      0
from          0
to            0
flightType    0
price         0
time          0
distance      0
agency        0
date          0
dtype: int64

Hotels Dataset - Missing Values
travelCode    0
userCode      0
name          0
place         0
days          0
price         0
total         0
date          0
dtype: int64


In [14]:
hotels.isnull().sum().sort_values(ascending=False).head(10)

travelCode    0
userCode      0
name          0
place         0
days          0
price         0
total         0
date          0
dtype: int64

In [17]:
print("Users Dataset - Duplicated Values")
print(users.duplicated().sum())

print("\nFlights Dataset - Duplicated Values")
print(flights.duplicated().sum())

print("\nHotels Dataset - Duplicated Values")
print(hotels.duplicated().sum())

Users Dataset - Duplicated Values
0

Flights Dataset - Duplicated Values
0

Hotels Dataset - Duplicated Values
0


In [19]:
users.describe()

,code,age
count,1340.000000,1340.000000
mean,669.500000,42.742537
std,386.968991,12.869779
min,0.000000,21.000000
25%,334.750000,32.000000
50%,669.500000,42.000000
75%,1004.250000,54.000000
max,1339.000000,65.000000


In [20]:
flights.describe()

,travelCode,userCode,price,time,distance
count,271888.000000,271888.000000,271888.00000,271888.000000,271888.000000
mean,67971.500000,667.505495,957.37503,1.421147,546.955535
std,39243.724665,389.523127,362.31189,0.542541,208.851288
min,0.000000,0.000000,301.51000,0.440000,168.220000
25%,33985.750000,326.000000,672.66000,1.040000,401.660000
50%,67971.500000,659.000000,904.00000,1.460000,562.140000
75%,101957.250000,1011.000000,1222.24000,1.760000,676.530000
max,135943.000000,1339.000000,1754.17000,2.440000,937.770000


In [21]:
hotels.describe()

,travelCode,userCode,days,price,total
count,40552.000000,40552.000000,40552.000000,40552.000000,40552.000000
mean,67911.794461,666.963726,2.499679,214.439554,536.229513
std,39408.199333,391.136794,1.119326,76.742305,319.331482
min,0.000000,0.000000,1.000000,60.390000,60.390000
25%,33696.750000,323.000000,1.000000,165.990000,247.620000
50%,67831.000000,658.000000,2.000000,242.880000,495.240000
75%,102211.250000,1013.000000,4.000000,263.410000,742.860000
max,135942.000000,1339.000000,4.000000,313.020000,1252.080000


### Observation

The Users dataset contains 1,340 records. User ages range from 21 to 65 years, with an average age of approximately 42.74 years. There are 5 companies and 3 gender categories in the dataset.

In [22]:
print("Users Dataset - Unique Values")
for col in users.columns:
    print(f"{col} : {users[col].nunique()}")    

print("\nFlights Dataset - Unique Values")
for col in flights.columns:
    print(f"{col} : {flights[col].nunique()}")    

print("\nHotels Dataset - Unique Values")
for col in hotels.columns:
    print(f"{col} : {hotels[col].nunique()}")       

Users Dataset - Unique Values
code : 1340
company : 5
name : 1338
gender : 3
age : 45

Flights Dataset - Unique Values
travelCode : 135944
userCode : 1335
from : 9
to : 9
flightType : 3
price : 490
time : 33
distance : 35
agency : 3
date : 999

Hotels Dataset - Unique Values
travelCode : 40552
userCode : 1310
name : 9
place : 9
days : 4
price : 9
total : 36
date : 199


### 1.3 Categorical Variables

In [38]:
print("Flight Types:")
print(flights["flightType"].unique())

print("\n Agencies")
print(flights["agency"].unique())

print("\n Origin cities:")
print(flights["from"].unique())

print("\n Destination cities:")
print(flights["to"].unique())

Flight Types:
<StringArray>
['firstClass', 'economic', 'premium']
Length: 3, dtype: str

 Agencies
<StringArray>
['FlyingDrops', 'CloudFy', 'Rainbow']
Length: 3, dtype: str

 Origin cities:
<StringArray>
['Recife (PE)', 'Florianopolis (SC)', 'Brasilia (DF)', 'Aracaju (SE)', 'Salvador (BH)', 'Campo Grande (MS)', 'Sao Paulo (SP)', 'Natal (RN)', 'Rio de Janeiro (RJ)']
Length: 9, dtype: str

 Destination cities:
<StringArray>
['Florianopolis (SC)', 'Recife (PE)', 'Brasilia (DF)', 'Salvador (BH)', 'Aracaju (SE)', 'Campo Grande (MS)', 'Sao Paulo (SP)', 'Natal (RN)', 'Rio de Janeiro (RJ)']
Length: 9, dtype: str


In [39]:
print("Hotel Names:")
print(hotels["name"].unique())

print("\n Hotel Cities:")
print(hotels["place"].unique())

Hotel Names:
<StringArray>
['Hotel A', 'Hotel K', 'Hotel BD', 'Hotel Z', 'Hotel AU', 'Hotel AF', 'Hotel BW', 'Hotel CB', 'Hotel BP']
Length: 9, dtype: str

 Hotel Cities:
<StringArray>
['Florianopolis (SC)', 'Salvador (BH)', 'Natal (RN)', 'Aracaju (SE)', 'Recife (PE)', 'Sao Paulo (SP)', 'Campo Grande (MS)', 'Rio de Janeiro (RJ)', 'Brasilia (DF)']
Length: 9, dtype: str


In [45]:
print("Genders:")
print(users["gender"].unique())

print("\n Companies:")
print(users["company"].unique())

print("\n No. of Companies:", users["company"].nunique())

Genders:
<StringArray>
['male', 'female', 'none']
Length: 3, dtype: str

 Companies:
<StringArray>
['4You', 'Monsters CYA', 'Wonka Company', 'Acme Factory', 'Umbrella LTDA']
Length: 5, dtype: str

 No. of Companies: 5


In [46]:
print("Flights date datatype:", flights["date"].dtype)
print("Hotels date datatype :", hotels["date"].dtype)

Flights date datatype: str
Hotels date datatype : str


In [47]:
print("Users:", users["code"].nunique())
print("Unique flight users:", flights["userCode"].nunique())
print("Unique hotel users:", hotels["userCode"].nunique())

Users: 1340
Unique flight users: 1335
Unique hotel users: 1310


In [48]:
flight_users_not_in_users = set(flights["userCode"]) - set(users["code"])

print("Flight users not found in Users dataset:",
      len(flight_users_not_in_users))

Flight users not found in Users dataset: 0


In [49]:
hotel_users_not_in_users = set(hotels["userCode"]) - set(users["code"])

print("Hotel users not found in Users dataset:",
      len(hotel_users_not_in_users))

Hotel users not found in Users dataset: 0


In [50]:
print("Unique Flight Travel Codes:",
      flights["travelCode"].nunique())

print("Unique Hotel Travel Codes:",
      hotels["travelCode"].nunique())

Unique Flight Travel Codes: 135944
Unique Hotel Travel Codes: 40552


In [51]:
flight_travel_codes = set(flights["travelCode"])
hotel_travel_codes = set(hotels["travelCode"])

print("Travel codes in Flights but not Hotels:",
      len(flight_travel_codes - hotel_travel_codes))

print("Travel codes in Hotels but not Flights:",
      len(hotel_travel_codes - flight_travel_codes))

Travel codes in Flights but not Hotels: 95392
Travel codes in Hotels but not Flights: 0


## 1.4 Dataset Relationships and Data Validation

In [52]:
# Check unique users across datasets

print("Users in Users dataset   :", users["code"].nunique())
print("Users in Flights dataset :", flights["userCode"].nunique())
print("Users in Hotels dataset  :", hotels["userCode"].nunique())

Users in Users dataset   : 1340
Users in Flights dataset : 1335
Users in Hotels dataset  : 1310


In [ ]:
# Checking whether all flight users exist in Users dataset

flight_users_not_in_users = set(flights["userCode"]) - set(users["code"])

print("Flight users not found in Users dataset:",
      len(flight_users_not_in_users))

Flight users not found in Users dataset: 0


In [54]:
# Checking whether all hotel users exist in Users dataset

hotel_users_not_in_users = set(hotels["userCode"]) - set(users["code"])

print("Hotel users not found in Users dataset:",
      len(hotel_users_not_in_users))

Hotel users not found in Users dataset: 0


In [55]:
# Checking travelCode relationship

flight_travel_codes = set(flights["travelCode"])
hotel_travel_codes = set(hotels["travelCode"])

print("Travel codes in Flights but not Hotels:",
      len(flight_travel_codes - hotel_travel_codes))

print("Travel codes in Hotels but not Flights:",
      len(hotel_travel_codes - flight_travel_codes))

Travel codes in Flights but not Hotels: 95392
Travel codes in Hotels but not Flights: 0


In [56]:
# Gender distribution

print(users["gender"].value_counts())

gender
male      452
female    448
none      440
Name: count, dtype: int64
